### ISSUE TITLE
Enable Encryption

### ISSUE THREAD

**Issue by danielgek** [couchbaselabs/nativescript-couchbase#45]
Hi, i'm trying to enable encryption in the plugin and as you can check in this [commit](https://github.com/NMA-SPMS/nativescript-couchbase/commit/04ae31ace2bdbb8943ba1ec3441794d68f4726eb) and i'm having some issues with ios runtime, so far i have updated the couchbase lite version and added sqlcipher lib too,
in the android side everything works ok, but in ios i can't get an instance of CBLDatabaseOptions

`Can't find variable: CBLDatabaseOptions`

`CBLDatabaseOptions` is in the `CBLManager.h:23`(couchbase headers files) 

btw i have deleted the static couchbase lite lib and added a PodFile with the 2 dependencies

And another question is the build.xcconfig, in the commit i removed the `-lsqlite3`, but when i'm usin the plugin in the ng-demo xcode adds it again, how can i prevent this ?

Thanks in advance, as soon as this is resolved i will make a pr with the changes

**Comment by nraboy** [comment]
Hey @danielgek,

I'm not much of an iOS wizard, but I'm going to include a few people who are.

In regards to the sqlcipher library, I'm wondering if it would be best to create a separate plugin for that.  It adds bloat to the application and most of the time people aren't going to want it included in their application unless they need it.  What do you think?

In any scenario, your contributions to this project are incredibly appreciated :-)

iOS contributors to this project below...

@hodgreeley @mehfuzh @NathanWalker

**Comment by danielgek** [comment]
Well, yes most of people will not use encryption so probably another plugin with encryption enable would make sense, but that would be replicating a huge part of this plugin, no ? Or the new plugin will only enable the encryption ? is that possible ?

**Comment by nraboy** [comment]
The beauty of TypeScript is that you can do inheritance and all that stuff you probably couldn't do with vanilla JavaScript, or at least not as easy.

It might be worth it to create an optional parameter in the Couchbase plugin when creating / opening the database.  The optional parameter could be that of your encryption class.  This would allow for separation I think.

What do you think?

**Comment by danielgek** [comment]
I like that approach, but how would you handle the additional sqlcypher lib and configs needed for iOS build.
Its possible to create a separate plugin with all that and with the Class that you can use on Couchbase(on creating/opening) or just do it directly in this plugin if you detect that parameter ?

**Comment by chimon2000** [comment]
I also see that couchbase lite now allows you to switch between [SQLite & ForestDB](https://www.couchbase.com/blog/2016/february/mobile-1.2-enable-forestdb) (which supports encryption by default). If ForestDB is favored by default here, then I think SQLCipher becomes unnecessary.

All that to say would it be better to allow developers to switch database implementations here, using DatabaseOptions, or in a separate plugin? My natural inclination is that configuration options are kind of integral, but I'm not sure how much of a size penalty those added dependencies bring.

**Comment by HodGreeley** [comment]
ForestDB is considered less mature/stable than SQLite currently.  I don't know the particular issues, but that's why it isn't the default.  I wouldn't want to pack in several databases just so I can choose on the fly.  Also, if it's done as a plugin, seems like that would make it more flexible if anyone wants to drop in another engine (although that may be a long shot).

**Comment by chimon2000** [comment]
@HodGreeley makes sense. I do like the flexibility of using plugins to manage engines.

**Comment by danielgek** [comment]
@nraboy i was able to implement the ios part, the documentation on [this page ](https://developer.couchbase.com/documentation/mobile/1.3/guides/couchbase-lite/native-api/database/index.html#enabling-encryption) is not updated i think, they suggest this implementation:
```objectivec
CBLDatabaseOptions* options = [[CBLDatabaseOptions alloc] init];
options.storageType = @"SQLite";
options.encryptionKey = @"password123456";
options.create = YES;
CBLDatabase* database = [manager openDatabaseNamed:@"db" withOptions:options error:nil];
```

but in [this wiki](https://github.com/couchbase/couchbase-lite-ios/wiki/Database-Encryption/7c7e11050662300e438ef40ef9d275516b5108b1#enabling-encryption) there is a implementation that actually works

```objectivec
CBLManager* mgr = [CBLManager sharedInstance];
[mgr registerEncryptionKey: @"password123456" forDatabaseNamed: @"my-db"];
CBLDatabase* db = [mgr databaseNamed: @"my-db" error: &error];
```
but it's not working yet, because i can't get the SQLCipher to link, so i get this error while trying to use it(witch is documented in the wiki):

```
Domain=CBLHTTP Code=501 "unimplemented" UserInfo={NSLocalizedFailureReason=unimplemented, NSLocalizedDescription=unimplemented}
```
you can check the progress [here](https://github.com/NMA-SPMS/nativescript-couchbase/commits/master)(i added one more page to ng-demo to test the database encryptions, and added a closing method too)

to resume the only missing par is to link the SQLCipher on the build

**Comment by danielgek** [comment]
@nraboy ok i got it to work with the changes i have made in those commits

Do you want me to submit a PR?

**Comment by nraboy** [comment]
Can you submit a PR with a descriptive summary on what you've done, and how you've tested? This thread got quite large and a summary would be helpful.

I'll do some testing on my end when you submit the PR.

Best,
